File: racket_viz/app/src/physics/racketGeometry.js
/**
 * Tennis-racket geometry -> mass distribution -> principal-axis inertia tensor.
 * Ported from physics/racket_geometry.py, using eigen3x3.js in place of
 * numpy.linalg.eigh (which already returns ascending eigenvalues with a
 * proper-rotation eigenvector matrix, so no separate det-flip step is needed
 * here -- eigh3x3 handles it internally).
 */
import { eigh3x3 } from "./eigen3x3.js";

/**
 * @param {object} [params]
 * @param {number} [params.handleLength=1.0]
 * @param {number} [params.hoopLength=2.0]
 * @param {number} [params.hoopWidth=1.0]
 * @param {number} [params.totalMass=1.0]
 * @param {number} [params.hoopFraction=0.60]
 * @param {number} [params.faceThickness=0.04]
 * @param {number} [params.numHandlePts=40]
 * @param {number} [params.numHoopPts=120]
 * @returns {{
 *   vertsBody: number[][], handleIdx: number[], hoopIdx: number[],
 *   I: number[], evecs: number[][], imin: number, imid: number, imax: number,
 *   faceThickness: number, faceNormalBody: number[], com: number[],
 *   comUnweighted: number[]
 * }}
 */
export function buildRacket(params = {}) {
  const {
    handleLength = 1.0,
    hoopLength = 2.0,
    hoopWidth = 1.0,
    totalMass = 1.0,
    hoopFraction = 0.6,
    faceThickness = 0.04,
    numHandlePts = 40,
    numHoopPts = 120,
  } = params;

  const handleFraction = 1.0 - hoopFraction;
  const hoopRadiusX = hoopLength / 2;
  const hoopRadiusY = hoopWidth / 2;
  const hoopCenterX = hoopRadiusX;

  const handlePts = [];
  for (let i = 0; i < numHandlePts; i++) {
    const x = -handleLength + (handleLength * i) / (numHandlePts - 1);
    handlePts.push([x, 0, 0]);
  }

  const hoopPts = [];
  for (let i = 0; i < numHoopPts; i++) {
    const theta = (2 * Math.PI * i) / numHoopPts;
    hoopPts.push([hoopCenterX + hoopRadiusX * Math.cos(theta), hoopRadiusY * Math.sin(theta), 0]);
  }

  const verticesRaw = [...handlePts, ...hoopPts];
  const handleIdx = Array.from({ length: numHandlePts }, (_, i) => i);
  const hoopIdx = Array.from({ length: numHoopPts }, (_, i) => i + numHandlePts);

  const masses = new Array(verticesRaw.length);
  const handleMassEach = (handleFraction * totalMass) / numHandlePts;
  const hoopMassEach = (hoopFraction * totalMass) / numHoopPts;
  for (const i of handleIdx) masses[i] = handleMassEach;
  for (const i of hoopIdx) masses[i] = hoopMassEach;

  const com = [0, 0, 0];
  for (let i = 0; i < verticesRaw.length; i++) {
    com[0] += verticesRaw[i][0] * masses[i];
    com[1] += verticesRaw[i][1] * masses[i];
    com[2] += verticesRaw[i][2] * masses[i];
  }
  com[0] /= totalMass;
  com[1] /= totalMass;
  com[2] /= totalMass;

  // Unweighted (purely geometric) centroid -- unlike `com`, this depends only
  // on the shape (handleLength/hoopLength/hoopWidth), never on hoopFraction
  // or totalMass, so it's a stable anchor for a live shape-only preview (see
  // main.js's previewRacketGeometry): `com` shifts along x by ~1 full unit as
  // hoopFraction alone is dragged across its range, which visibly translates
  // the racket on screen even though its actual shape hasn't changed.
  const comUnweighted = [0, 0, 0];
  for (const v of verticesRaw) {
    comUnweighted[0] += v[0];
    comUnweighted[1] += v[1];
    comUnweighted[2] += v[2];
  }
  comUnweighted[0] /= verticesRaw.length;
  comUnweighted[1] /= verticesRaw.length;
  comUnweighted[2] /= verticesRaw.length;

  const vertsC = verticesRaw.map((v) => [v[0] - com[0], v[1] - com[1], v[2] - com[2]]);

  const Itensor = [
    [0, 0, 0],
    [0, 0, 0],
    [0, 0, 0],
  ];
  for (let i = 0; i < vertsC.length; i++) {
    const r = vertsC[i];
    const m = masses[i];
    const r2 = r[0] * r[0] + r[1] * r[1] + r[2] * r[2];
    for (let a = 0; a < 3; a++) {
      for (let b = 0; b < 3; b++) {
        Itensor[a][b] += m * ((a === b ? r2 : 0) - r[a] * r[b]);
      }
    }
  }

  const { values: I, vectors: evecs } = eigh3x3(Itensor);

  const vertsBody = vertsC.map((r) => [
    r[0] * evecs[0][0] + r[1] * evecs[1][0] + r[2] * evecs[2][0],
    r[0] * evecs[0][1] + r[1] * evecs[1][1] + r[2] * evecs[2][1],
    r[0] * evecs[0][2] + r[1] * evecs[1][2] + r[2] * evecs[2][2],
  ]);

  // face_normal_body = evecs^T @ [0,0,1] = the 3rd ROW of evecs (columns are
  // eigenvectors, so evecs^T's 3rd row is evecs's 3rd row transposed back)
  const rawNormal = [evecs[2][0], evecs[2][1], evecs[2][2]];
  const normalMag = Math.sqrt(rawNormal[0] ** 2 + rawNormal[1] ** 2 + rawNormal[2] ** 2);
  const faceNormalBody = rawNormal.map((x) => x / normalMag);

  return {
    vertsBody,
    handleIdx,
    hoopIdx,
    I,
    evecs,
    imin: 0,
    imid: 1,
    imax: 2,
    faceThickness,
    faceNormalBody,
    com,
    comUnweighted,
  };
}

File: racket_viz/physics/racket_geometry.py
"""Tennis-racket geometry → mass distribution → principal-axis inertia tensor.

Adapted from `summer-2026/dzhanibekov_display.py` (Sections 2 & 3): a handle segment
plus an elliptical hoop, uniformly-massed points on each, inertia tensor computed
about the center of mass and diagonalized to find the principal axes.
"""
import numpy as np


def build_racket(
    handle_length=1.0,
    hoop_length=2.0,
    hoop_width=1.0,
    total_mass=1.0,
    hoop_fraction=0.60,
    face_thickness=0.04,
    num_handle_pts=40,
    num_hoop_pts=120,
):
    """Build racket geometry and its principal-axis inertia tensor.

    Returns a dict with:
      verts_body   : (N, 3) vertex positions in the principal-axis (body) frame
      handle_idx   : indices into verts_body for the handle
      hoop_idx     : indices into verts_body for the hoop
      I            : (I1, I2, I3), I1 < I2 < I3
      evecs        : (3, 3) columns are principal axes in the original geometric frame
      imin,imid,imax : int, indices 0/1/2 identifying the stable/unstable/stable axes
      face_thickness : float, passed through for rendering the two racket faces
      face_normal_body : (3,) unit normal to the racket face, in the body frame --
        exactly the geometric [0,0,1] normal re-expressed in the principal-axis
        basis. Guaranteed to equal the imax axis (up to sign): the racket is
        planar (every vertex has z=0 before the principal-axis rotation), and
        for any planar mass distribution the out-of-plane axis is automatically
        a principal axis with the *largest* moment (perpendicular axis theorem:
        I_z = I_x + I_y >= max(I_x, I_y)).
    """
    handle_fraction = 1.0 - hoop_fraction
    hoop_radius_x = hoop_length / 2
    hoop_radius_y = hoop_width / 2
    hoop_center_x = hoop_radius_x

    handle_x = np.linspace(-handle_length, 0.0, num_handle_pts)
    handle_points = np.column_stack([
        handle_x, np.zeros(num_handle_pts), np.zeros(num_handle_pts),
    ])

    theta = np.linspace(0, 2 * np.pi, num_hoop_pts, endpoint=False)
    hoop_points = np.column_stack([
        hoop_center_x + hoop_radius_x * np.cos(theta),
        hoop_radius_y * np.sin(theta),
        np.zeros(num_hoop_pts),
    ])

    vertices_raw = np.vstack([handle_points, hoop_points])
    handle_idx = np.arange(num_handle_pts)
    hoop_idx = np.arange(num_handle_pts, num_handle_pts + num_hoop_pts)

    masses = np.zeros(len(vertices_raw))
    masses[handle_idx] = (handle_fraction * total_mass) / len(handle_idx)
    masses[hoop_idx] = (hoop_fraction * total_mass) / len(hoop_idx)

    com = np.sum(vertices_raw * masses[:, None], axis=0) / total_mass
    verts_c = vertices_raw - com

    I_tensor = np.zeros((3, 3))
    for r, m in zip(verts_c, masses):
        r2 = np.dot(r, r)
        I_tensor += m * (r2 * np.eye(3) - np.outer(r, r))

    evals, evecs = np.linalg.eigh(I_tensor)
    order = np.argsort(evals)
    evals = evals[order]
    evecs = evecs[:, order]

    # eigh's sign/handedness is arbitrary; flip the last axis if needed so the
    # principal frame is a proper (det=+1) rotation of the geometric frame.
    if np.linalg.det(evecs) < 0:
        evecs[:, -1] *= -1

    verts_body = verts_c @ evecs
    face_normal_body = evecs.T @ np.array([0.0, 0.0, 1.0])
    face_normal_body /= np.linalg.norm(face_normal_body)

    return {
        "verts_body": verts_body,
        "handle_idx": handle_idx,
        "hoop_idx": hoop_idx,
        "I": evals,
        "evecs": evecs,
        "imin": 0,
        "imid": 1,
        "imax": 2,
        "face_thickness": face_thickness,
        "face_normal_body": face_normal_body,
        "com": com,
    }
